flang/lib/Optimizer/Analysis/AliasAnalysis.cpp

```diff
@@ -122,13 +122,32 @@ AliasResult AliasAnalysis::alias(Value lhs, Value rhs) {
   }
 
   if (lhsSrc.kind == rhsSrc.kind) {
+    // If the kinds and origins are the same, then lhs and rhs must alias unless
+    // either source is approximate.  Approximate sources are for parts of the
+    // origin, but we don't have info here on which parts and whether they
+    // overlap, so we normally return MayAlias in that case.
+    //
+    // There is an exceptional case: The origins might compare unequal because
+    // only one has !isData().  If that source is approximate and the other is
+    // not, then the former is the source for the address *of* a pointer
+    // component in a composite, and the latter is for the address of that
+    // composite.  As for the address of any composite vs. the address of one of
+    // its components, a store to one can affect a load from the other, so the
+    // result is MayAlias.
     if (lhsSrc.origin == rhsSrc.origin) {
       LLVM_DEBUG(llvm::dbgs()
                  << "  aliasing because same source kind and origin\n");
       if (approximateSource)
         return AliasResult::MayAlias;
       return AliasResult::MustAlias;
     }
+    if (lhsSrc.origin.u == rhsSrc.origin.u &&
+        ((lhsSrc.approximateSource && !lhsSrc.isData() && !rhsSrc.approximateSource) ||
+         (rhsSrc.approximateSource && !rhsSrc.isData() && !lhsSrc.approximateSource))) {
+      LLVM_DEBUG(llvm::dbgs()
+                 << "  aliasing between composite and pointer component\n");
+      return AliasResult::MayAlias;
+    }
```

Comment:
@Renaud-K posted the following [comment as part of a commit](https://github.com/jdenny-ornl/llvm-project/commit/484cb901cb3f5ae934b55629bb8537e42d3a832d#r142575988), and I'm trying to move the conversation into this PR:

> I don't think `approximateSource` is a reliable way of meaning `component`. We could have an optimized out `fir.ccordinate_of %memref %c0`, still referring to the same source.

Reply:
@Renaud-K I could extend `Source` with a field like `bool component;` It would be set to true for any `DesignateOp` specifying a component, and `approximateSource` would also still be set to true then.  Does that seem like a reasonable direction to try?

Reply:
`DesignateOp` is later lowered to `coordinate_of` for instance. The absence of which does not mean we are not dealing with a component. The alias analysis has to work at all levels.  Again, `fir.coordinate_of %memref %c0` is the same as `%memref`. We need to provide consistent results whether `fir.coordinate_of` is present or at least error on the conservative side.

Reply:
> `DesignateOp` is later lowered to `coordinate_of` for instance.

Thanks.  In a simple example, I see that

```
%10 = hlfir.designate %1#0{"p"}   {fortran_attrs = #fir.var_attrs<pointer>} : (!fir.ref<!fir.type<_QMmTt{p:!fir.box<!fir.ptr<f32>>}>>) -> !fir.ref<!fir.box<!fir.ptr<f32>>>
```

becomes

```
%10 = fir.field_index p, !fir.type<_QMmTt{p:!fir.box<!fir.ptr<f32>>}>
%11 = fir.coordinate_of %1, %10 : (!fir.ref<!fir.type<_QMmTt{p:!fir.box<!fir.ptr<f32>>}>>, !fir.field) -> !fir.ref<!fir.box<!fir.ptr<f32>>>
```

The latter pattern seems simple enough to detect.  I did not find a later version of that IR where `fir.coordinate_of` appeared but `!fir.field` and `fir.field_index` had been optimized away.  Is your `fir.coordinate_of %memref %c0` an example of that?

> The absence of which does not mean we are not dealing with a component.

Is it true that either `hlfir.designate` or `fir.coordinate_of` must be present when accessing a component?  If not, then, independently of this PR, we risk `approximateSource=false` and an invalid MustAlias, right?

Reply:
MustAlias is correct if we can determine the source more accurately, and MayAlias is a conservative error. But in your case we could go from MayAlias (because we detect components) to NoAlias which is too optimistic. Maybe we can keep using the type, such as, if the type of the source is a derived type and the type of the starting point is not then we have a component.

Reply:
This is almost what you suggested:

```
     if (lhsSrc.origin.u == rhsSrc.origin.u &&
-        ((lhsSrc.approximateSource && !lhsSrc.isData() && !rhsSrc.approximateSource) ||
-         (rhsSrc.approximateSource && !rhsSrc.isData() && !lhsSrc.approximateSource))) {
+        ((lhsSrc.isRecordWithPointerComponent() && !rhsSrc.isData()) ||
+         (rhsSrc.isRecordWithPointerComponent() && !lhsSrc.isData()))) {
       LLVM_DEBUG(llvm::dbgs()
                  << "  aliasing between composite and non-data component\n");
       return AliasResult::MayAlias;
     }
```

Reply:
Yes. Thank you. Now from a testing stand point I am not sure if we explicitly test for the case of component pointers **data**. Maybe they will be handled by the general processing of pointers and targets but it would be nice to check for aliasing since they should be aliasing with other target or pointers. Do you agree with this statement?

Reply:
> Now from a testing stand point I am not sure if we explicitly test for the case of component pointers **data**.

That requires a load to extract the address in the pointer.  That's the address of the data.  However, getSource doesn't handle load on, for example, an hlfir.designate yet, so we get an Indirect and thus a MayAlias.  Is that what you mean?

I am working on a patch to extend load handling to that case.  The general pointer component logic confused my testing efforts, so I wanted to address it first.

Reply:
Hmm, we do handle loads as long as they are not dynamic. If we cannot statically determine what is loaded (the comment calls it a leaf source), then it is indirect. There is not much we can do about it. What is not yet handled is the case when the leaf source is a SourceKind::Allocate but that should not prevent you from working with pointers and dummy arguments.

Reply:
%11 below is the address of the data in `x%p` where `p` is a pointer:

```
  func.func @_QMmPfoo() {
    ...
    %2 = fir.alloca !fir.type<_QMmTt{p:!fir.box<!fir.ptr<i32>>}> {bindc_name = "x", uniq_name = "_QMmFfooEx"}
    %3:2 = hlfir.declare %2 {uniq_name = "_QMmFfooEx"} : (!fir.ref<!fir.type<_QMmTt{p:!fir.box<!fir.ptr<i32>>}>>) -> (!fir.ref<!fir.type<_QMmTt{p:!fir.box<!fir.ptr<i32>>}>>, !fir.ref<!fir.type<_QMmTt{p:!fir.box<!fir.ptr<i32>>}>>)
    ...
    %9 = hlfir.designate %3#0{"p"}   {fortran_attrs = #fir.var_attrs<pointer>} : (!fir.ref<!fir.type<_QMmTt{p:!fir.box<!fir.ptr<i32>>}>>) -> !fir.ref<!fir.box<!fir.ptr<i32>>>
    %10 = fir.load %9 : !fir.ref<!fir.box<!fir.ptr<i32>>>
    %11 = fir.box_addr %10 : (!fir.box<!fir.ptr<i32>>) -> !fir.ptr<i32>
    ...
    return
  }
```

The LoadOp case in getSource needs to be extended to handle the hlfir.designate.  My current draft of that patch does it in getOriginalDef.

I also have a patch to extend LoadOp for alloca.

Reply:
I am not sure we can do anything about `hlfir.designate` in `getOriginalDef`. The value loaded from %9 could be anything. We cannot follow it to %3. A load from %3 has nothing to do with a load from %9.

Reply:
%11 below is the address of the data in p, which is a pointer:

```
  func.func @_QMmPfoo() {
    ...
    %2 = fir.alloca !fir.box<!fir.ptr<i32>> {bindc_name = "p", uniq_name = "_QMmFfooEp"}
    %3:2 = hlfir.declare %2 {fortran_attrs = #fir.var_attrs<pointer>, uniq_name = "_QMmFfooEp"} : (!fir.ref<!fir.box<!fir.ptr<i32>>>) -> (!fir.ref<!fir.box<!fir.ptr<i32>>>, !fir.ref<!fir.box<!fir.ptr<i32>>>)
   ...
    %10 = fir.load %3#0 : !fir.ref<!fir.box<!fir.ptr<i32>>>
    %11 = fir.box_addr %10 : (!fir.box<!fir.ptr<i32>>) -> !fir.ptr<i32>
    ...
    return
  }
```

Isn't this the same as the previous example except that, without the designate, we're loading all of %2 instead of only one component of it?

Reply:
To finish the thought, I don't see why we have to give up (SourceKind=Indirect forces MayAlias) on tracking aliasing for a pointer just because it's a member of an alloca rather than the entire alloca.

Reply:
My concern would be to modify getOriginalDef to return anything other than Source::Indirect. 
`%11` with the designate case is always indirect. We have no idea what was stored in `%9`, it could a global, an allocatable or a dummy: we do not know.
 
There is room to improve on SourceKind::Indirect, such as using the type system to distinguish !fir.ptr<T> from other memory references but that would be a different discussion.  In the example just above, `%11` would then be indirect too since we are following the data and not interested in the box address which happens to be an SourceKind::Alloca. Today, I would guess, we would get SourceKind::Alloca with an isData flag but it would be better modeled with SourceKind::Indirect and isData. Then we could do with SourceKind::Indirect what we used to do clumsily with SourceKind::Direct. 

But I think, we should take it one step at a time.

Reply:
> My concern would be to modify getOriginalDef to return anything other than Source::Indirect.

For the record, getOriginalDef does not itself choose a SourceKind.  Moreover, after getSource's LoadOp case calls getOriginalDef, getSource currently returns any of Indirect, Argument, or Global.  That is true independently of this patch.  No modification required.

(As you have noted, there are cases like alloca that getSource's LoadOp case doesn't currently fully handle.  As a result, it currently returns Indirect sometimes where, for consistency with the current handling of dummy args and globals, it should return Allocate.)

> `%11` with the designate case is always indirect. We have no idea what was stored in `%9`, it could a global, an allocatable or a dummy: we do not know.

In that example, what was stored in %9 is the address of the pointer. In my above example without hlfir.designate, %3 is the address of the pointer.  As far as I understand, the difference between those is whether that address is at a base address (for a global, dummy arg, alloca, etc.) or at an offset from that base address.

I don't see why that difference *should* affect the SourceKind for any value that is computed from the address of the pointer.  In particular, if the source origin (i.e., %2) were a dummy arg or global, the SourceKind for %11 in my above example without hflir.designate would currently be Argument or Global, but the SourceKind for %11 in my above example with hlfir.designate would currently be Indirect.  I see no reason for that difference.  In a later patch, I would like to fix that.  (Again, I would also like to add support for the alloca/Allocate case shown in the examples.)

> There is room to improve on SourceKind::Indirect, such as using the type system to distinguish !fir.ptr from other memory references but that would be a different discussion. In the example just above, `%11` would then be indirect too since we are following the data and not interested in the box address which happens to be an SourceKind::Alloca. Today, I would guess, we would get SourceKind::Alloca with an isData flag but it would be better modeled with SourceKind::Indirect and isData. Then we could do with SourceKind::Indirect what we used to do clumsily with SourceKind::Direct.

Whatever we decide to call the SourceKind for that case, do we agree that hlfir.designate should not change the SourceKind for %11 between my previous two examples?

> But I think, we should take it one step at a time.

Agreed.  Hopefully the current patch I'm trying to develop can be a next step.

> It would impact @tblah who is handling the (global, isData) separately. These two would become (indirect, isData)

Where is the discussion of his work?

Reply:
@jdenny-ornl my work was proposed here https://discourse.llvm.org/t/rfc-propagate-fir-alias-analysis-information-using-tbaa/73755

There have since been changes since then to add support for Direct variables: https://github.com/llvm/llvm-project/commit/6242c8ca18bcd6765094c73dd2c8b49200a6cec8

I think in this case the proposed changes won't alter the TBAA tags. My pass doesn't tag SourceKind::Alloca (by default) or SourceKind::Indirect (at all) (there are still TBAA tags added later during codegen to distinguish box metadata access from data access, but that is different again).

Reply:
@tblah Thanks for the pointers!